velox/expression/Expr.cpp

```diff
@@ -420,14 +420,17 @@ bool Expr::checkGetSharedSubexprValues(
     assert(missingRows); // lint
     missingRows->deselect(*sharedSubexprRows_);
 
-    // Fix finalSelection at "rows" if missingRows is a strict subset to avoid
-    // losing values outside of missingRows.
-    bool updateFinalSelection = context.isFinalSelection() &&
-        (missingRows->countSelected() < rows.countSelected());
-    VarSetter finalSelectionOr(
-        context.mutableFinalSelection(), &rows, updateFinalSelection);
-    VarSetter isFinalSelectionOr(
-        context.mutableIsFinalSelection(), false, updateFinalSelection);
+    // Add the missingRows to sharedSubexprRows_ that will eventually be
+    // evaluated and added to sharedSubexprValues_.
+    sharedSubexprRows_->select(*missingRows);
+    // Fix finalSelection to the updated sharedSubexprRows_ to avoid
+    // losing values outside missingRows.
+    VarSetter finalSelectionPreservePrecomputedValues(
```

Comment:
I wonder if it is possible for sharedSubexprRows_ to have rows that are not present in context->finalSelection(). I think this may happen if some rows are needed for evaluating one expression, but not another. To cover this case we may need to set finalSelection to a union of finalSelection and sharedSubexprRows_.

Reply:
On a side note: I was also looking through the implementation in evalWithMemo() to make sure we dont have a similar bug there as it has similar functionality of re-using results. Over there the same logic exists that checks for 

```
bool updateFinalSelection = context.isFinalSelection() &&
        (missingRows->countSelected() < rows.countSelected());
```

However, it works there because the eval() is called on the 'result' and not the cached vector (dictionaryCache_) so we ever only need to worry about preserving 'rows' or 'finalSelection' (if set). However, in this case we call eval() on the shared result vector (sharedSubexprRows_) and hence need to ensure that the union of (sharedSubexprRows_ & rows/finalSelection) is preserved.

Reply:
> > Just to confirm, do you mean test coverage with respect to adding a case where final selection is not a superset of sharedSubExpressionRows?
> 
> I'm thinking that we might be missing test cases that would cover the following scenario:
> 
> ```
> I wonder if it is possible for sharedSubexprRows_ to have rows that are not present in context->finalSelection(). I think this may happen if some rows are needed for evaluating one expression, but not another. To cover this case we may need to set finalSelection to a union of finalSelection and sharedSubexprRows_.
> ```

While writing the test for this as you suggested earlier (see quoted reply above), I took another look at how the evalWithMemo is implemented and my comment above. I feel like the flow in evalWithMemo is more seamless and less confusing and it also avoids overriding finalSelection in an unintuitive way (union of sharedRows + finalSelection). I think it would be a good idea to change the flow for checkGetSharedSubexprValues() to mimic what we do in evalWithMemo and make them more uniform and consistent.
The flow goes as follows.
**Let,**
input rows = 'rows'
Output values = 'outputValues'
finalSelection = 'finalSelectionRows'
sharedSubExprRows or memoized rows = 'sharedRows'
sharedSubExprRows or memoized values= 'sharedValues'

	1. Calculate cachedRows = Rows ∩ sharedRows
	2. Copy cachedRows from sharedValues to outputValues
	3. Calculate uncachedRows = Rows - sharedRows
	4. If uncachedRows are not empty
		a. Set finalSelection to 'rows' **IF**
			i. uncachedRows is a strict subset of rows
			ii. not already set (since 'rows' can ever only be a subset of finalSelection, this is safe to do)
		b. Call eval(uncachedRows , outputValues)
		c. Copy newly evaluated uncachedRows from outputValues to sharedValues
		d. Update sharedRows to include newly added uncachedRows
	5. Return outputValues that now have 'row' set

@mbasmanova what do you think?

Reply:
> since 'rows' can ever only be a subset of finalSelection

I was under assumption that this is not the case for CSE shared across multiple expressions.

Reply:
Since in this proposed flow we pass 'outputValues' to eval() instead of 'sharedValues' we are only concerned about preserving rows in 'finalSelection' corresponding to the expression tree that we are currently evaluating.
this also makes sure that logic related to manipulating the vector in sharedValues is limited to the checkGetSharedSubexprValues() function and we will not run into issues of overwriting like we did in this bug